sgf/scanner.go

```diff
@@ -0,0 +1,160 @@
+package sgf
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+)
+
+const (
+	Eof = iota
+	Whitespace
+	LeftParen
+	RightParen
+	LeftBracket
+	RightBracket
+	Semicolon
+	Backslash
+	String
+)
+
+var eof = rune(0)
+
+// isWhiteSpace is a check to see if a rune is whitespace (space, tab, newline, carriage return)
+func isWhitespace(ch rune) bool {
+	return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r'
+}
+
+// isSpecial is a check for special characters used in SGF files
+func isSpecial(ch rune) bool {
+	return ch == '(' || ch == ')' || ch == '[' || ch == ']' || ch == ';'
+}
+
+// The scanner will "tokenize" the input into Token structs
+type Token struct {
+	Type int
+	Raw  string
+}
+
+// Scanner wraps bufio.Reader but has the public Scan() function to scan tokens
+type Scanner struct {
+	*bufio.Reader
+}
+
+// Creates a new scanner object out of an io.Reader interface
+func NewScanner(r io.Reader) *Scanner {
+	return &Scanner{bufio.NewReader(r)}
+}
+
+// Either return the current rune or EOF (null)
+func (s *Scanner) read() rune {
+	ch, _, err := s.ReadRune()
+	if err != nil {
+		return eof
+	}
+	return ch
+}
+
+// Go back one rune
+func (s *Scanner) unread() {
+	s.UnreadRune()
+}
+
+// Scan is the function that tokenizes the input and outputs tokens
+func (s *Scanner) Scan() *Token {
+	// read in a character
+	ch := s.read()
+
+	// if it's whitespace
+	if isWhitespace(ch) {
+		// then go back and scan as much whitespace as possible
+		s.unread()
+		return s.scanWhitespace()
+		// if it's not a special character or EOF, then it's a "string"
+	} else if !isSpecial(ch) && ch != eof {
+		// go back and scan as much of the string as possible
+		s.unread()
+		return s.scanString()
+	}
+
+	// otherwise, handle the special characters
+	switch ch {
+	case '(':
+		return &Token{
+			Type: LeftParen,
+			Raw:  "(",
+		}
+	case ')':
+		return &Token{
+			Type: RightParen,
+			Raw:  ")",
+		}
+	case '[':
+		return &Token{
+			Type: LeftBracket,
+			Raw:  "[",
+		}
+	case ']':
+		return &Token{
+			Type: RightBracket,
+			Raw:  "]",
+		}
+	case ';':
+		return &Token{
+			Type: Semicolon,
+			Raw:  ";",
+		}
+	default:
```

Comment:
Actually, wouldn't this be an error/unexpected token?

Reply:
Agreed, fixed